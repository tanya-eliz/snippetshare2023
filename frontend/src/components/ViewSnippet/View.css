.view-body{
    padding: 3%;
    padding-top:2.5%;
}

.content-body{
    margin: 1% 0 1% 0 !important;
}

.view-count{
    display:flex;
    flex-direction: row;
    justify-content: flex-end;
    align-items: center;
}

.copy-button{
    display:flex;
    flex-direction: row;
    justify-content: flex-end;
    align-items: center;
}