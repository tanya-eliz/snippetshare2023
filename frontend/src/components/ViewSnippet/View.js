import {useState, useEffect} from 'react';
import useFetch from "./useFetch";
import {useParams} from 'react-router-dom';
import Grid from '@mui/material/Grid';
import ViewIcon from '../../assets/visible.png'
import './View.css';
import TextField from '@mui/material/TextField';
import CopyToClipboardButton from '../../atoms/Copy/Copy';

export const View = () => {
    const {id} = useParams();
    const {data: snippets, isPending, error} = useFetch('http://localhost:4000/api/snippets/' + id);
    const [seconds, setSeconds] = useState(0);
    const [countDown, setCountDown] = useState(null);

    useEffect(() => {
        if (!isPending && snippets && snippets.expiry) {
            setSeconds(getTimeLeftTillExpiry(snippets.expiry));
        }
    }, [snippets, isPending]);

    useEffect(() => {
        
        if (seconds < 0) {
            clearInterval(countDown);
        } else if (countDown === null && seconds >= 0) {
            setCountDown(setInterval(() => {
                setSeconds((second) => second - 1 );
            }, 1000));
        }
    }, [seconds, countDown]);

    const getTimeLeftTillExpiry = (expiry) => {
        const now = new Date();
        const expiryDate = new Date(expiry);
        const diff = expiryDate.getTime() - now.getTime();
        const seconds = Math.floor(diff / 1000);
        console.log(expiryDate, seconds)
        return seconds;
    }

    const renderCountdown = (seconds) => {
        const days = Math.floor(seconds / (3600 * 24));
        const hours = Math.floor(seconds % (3600 * 24) / 3600);
        const minutes = Math.floor(seconds % 3600 / 60);
        const secs = Math.floor(seconds % 60);

        if (seconds >= 0) {
            return  <div>Expiring in {days} days, {hours} hours, {minutes} minutes, {secs} seconds</div>
        } else {
            return <div>Expired</div>
        }
    }

    return (
        <div>
            {
                isPending ? 
                <div>Loading...</div> : 
                error ? <div>{error}</div> :
                snippets && (
                    <div className="view-body">
                        <Grid container spacing={0} className="view-header">
                            <Grid item xs={10}>
                                <h1>{snippets.title}</h1>
                            </Grid>
                            <Grid item xs={2}>
                                <span className="view-count">{snippets.viewCount}<img src={ViewIcon} alt="view"/></span>
                            </Grid>
                            <Grid item xs={12}>
                                <p style={{fontStyle: 'Italic'}}>{snippets.expiry? renderCountdown(seconds) : <div>Never</div>}</p>
                            </Grid>
                        </Grid>

                            <Grid container spacing={0} className="content-body">
                                <Grid item xs={10}>
                                    Content
                                </Grid>
                                <Grid item xs={2} className="copy-button">
                                    <CopyToClipboardButton textCopy={snippets.content}/>
                                </Grid>
                            </Grid>

                            <TextField
                                className="content-body"
                                multiline
                                fullWidth
                                minRows = {12}
                                maxRows = {30}
                                value={snippets.expiry && seconds>0 ? snippets.content : 'expired'}
                                InputProps={{readOnly: true}}
                            />
                    </div>
                )
            }
        </div>
    )
}