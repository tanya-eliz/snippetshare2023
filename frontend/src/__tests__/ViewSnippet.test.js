/* eslint-disable testing-library/no-container */
/* eslint-disable testing-library/await-async-query */
import {render, screen, cleanup} from "@testing-library/react";
import "@testing-library/jest-dom";
import View from "../components/ViewSnippet/View";

afterEach(() => {
    cleanup();
});

test("should render View component", () => {
    render(<View />);
    const viewElem = screen.getByTestId("view");
    expect(viewElem).toBeInTheDocument();
});

test("View component matches snapshot", () => {
    const { asFragment } = render(<View />);
    expect(asFragment()).toMatchSnapshot();
});
