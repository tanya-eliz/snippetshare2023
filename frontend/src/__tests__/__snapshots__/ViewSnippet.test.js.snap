// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`View component matches snapshot 1`] = `
<DocumentFragment>
  <div
    data-testid="view"
  >
    <div>
      Loading...
    </div>
  </div>
</DocumentFragment>
`;
